p4-14/v1.0.5/tex/p4.pt

```diff
@@ -705,6 +700,22 @@ References at run time to a header instance (or one of its fields) which is
 not valid results in a special ``undefined'' value.  The implications of this 
 depend on the context.
 \end{itemize}
+    
+To help programers write parsers for header stacks, P4 provides two
+keywords, \texttt{next} and \texttt{last}, that may be used to refer
+to header instances within a stack. These references are advanced
+automatically as the parser invokes \texttt{extract} in the parser. It
+is illegal to use these keywords outside of a parser.
+    
+More specifically, the keyword \texttt{next} refers to the header
+instance at index \texttt{0} initially, and is advanced each time that
+\texttt{extract} is invoked on an instance within the stack. If
```

Comment:
If it is intended to allow extract on a constant index within a header stack (which I believe Chris Dodd mentioned there are extant uses of in P4_14 programs, e.g. for 2-entry header stacks of VLAN tag headers), then perhaps that should be explicitly mentioned somewhere as permitted?

Also, perhaps the "is advanced each time" condition should be qualified a bit more, e.g. "is advanced each time that \texttt{extract} is invoked on the [next] element of a header stack."

Reply:
Good suggestions both.

Added the paragraph:
```
Note that while it is legal to extract to a constant index in a header
stack and extract to the element referred by \texttt{next}, it is not
recommended. It will likely be difficult for programmers to keep track
of which element in the header stack \texttt{next} refers to when
operations are combined in this way.
```

Clarified by adding:
```
and is advanced each time that \texttt{extract} is invoked on any instance within the stack. 
```